src/Mod/Sketcher/Gui/CommandConstraints.cpp

```diff
@@ -990,6 +990,1165 @@ void CmdSketcherConstraint::activated(int /*iMsg*/)
+        }
+        setCursor(cursorPixmap, hotX, hotY, false);
+    }
+
+    void deactivated() override
+    {
+        Gui::Command::abortCommand();
+        //TODO: When user exit sketch edit while tool is activated solve() crashes. solve is needed to refresh the sketch after abortCommand
+        Obj->solve();
+        sketchgui->draw(false, false); // Redraw
+    }
+
+    virtual void registerPressedKey(bool pressed, int key)
+    {
+        if ((key == SoKeyboardEvent::RIGHT_SHIFT || key == SoKeyboardEvent::LEFT_SHIFT) && pressed) {
+            if (availableConstraint == AvailableConstraint::FIRST) {
+                availableConstraint = AvailableConstraint::SECOND;
+            }
+            else if (availableConstraint == AvailableConstraint::SECOND) {
+                availableConstraint = AvailableConstraint::THIRD;
+            }
+            else if (availableConstraint == AvailableConstraint::THIRD) {
+                availableConstraint = AvailableConstraint::FOURTH;
+            }
+            else if (availableConstraint == AvailableConstraint::FOURTH) {
+                availableConstraint = AvailableConstraint::FIFTH;
+            }
+            else if (availableConstraint == AvailableConstraint::FIFTH || availableConstraint == AvailableConstraint::RESET) {
+                availableConstraint = AvailableConstraint::FIRST;
+            }
+            makeAppropriateConstraint(previousOnSketchPos);
+        }
+    }
+
+    virtual void mouseMove(Base::Vector2d onSketchPos)
+    {
+        const std::vector<Sketcher::Constraint*>& ConStr = Obj->Constraints.getValues();
+        previousOnSketchPos = onSketchPos;
+
+        //Change distance constraint based on position of mouse.
+        if (specialConstraint == SpecialConstraint::LineOr2PointsDistance)
+            updateDistanceType(onSketchPos);
+
+        //Move constraints
+        if (numberOfConstraintsCreated > 0) {
+            bool oneMoved = false;
+            for (int i = 0; i < numberOfConstraintsCreated; i++) {
+                if (ConStr[ConStr.size() - 1 - i]->isDimensional()) {
+                    Base::Vector2d pointWhereToMove = onSketchPos;
+
+                    if (specialConstraint == SpecialConstraint::Block) {
+                        if (i == 0)
+                            pointWhereToMove.x = Obj->getPoint(selPoints[0].GeoId, selPoints[0].PosId).x;
+                        else
+                            pointWhereToMove.y = Obj->getPoint(selPoints[0].GeoId, selPoints[0].PosId).y;
+                    }
+                    moveConstraint(ConStr.size() - 1 - i, pointWhereToMove);
+                    oneMoved = true;
+                }
+            }
+            if (oneMoved)
+                sketchgui->draw(false, false); // Redraw
+        }
+    }
+
+    virtual bool pressButton(Base::Vector2d onSketchPos)
+    {
+        return true;
+    }
+
+    virtual bool releaseButton(Base::Vector2d onSketchPos)
+    {
```

Comment:
It doesn't harm to refactor this function.

Reply:
Please let me know if it is now ok.

Reply:
Much better now, thanks!